test/unit/services/CustomsNotificationRetryServiceSpec.scala

```diff
@@ -124,6 +124,29 @@ class CustomsNotificationRetryServiceSpec extends UnitSpec with MockitoSugar wit
         eventually(verify(mockNotificationWorkItemRepo, times(0)).setCompletedStatus(any[BSONObjectID], any[ResultStatus]))
         infoLogVerifier("Existing permanently failed notifications found for client id: ClientId. Setting notification to permanently failed")
       }
+
+      "returned HasSaved is true BEFORE push has succeeded" in {
+        when(mockNotificationWorkItemRepo.permanentlyFailedByClientIdExists(NotificationWorkItemWithMetricsTime1.clientId)).thenReturn(Future.successful(false))
+        when(mockNotificationWorkItemRepo.saveWithLock(refEq(NotificationWorkItemWithMetricsTime1), refEq(InProgress))).thenReturn(Future.successful(WorkItem1))
+        when(mockNotificationWorkItemRepo.setCompletedStatus(WorkItem1.id, PermanentlyFailed)).thenReturn(Future.successful(()))
+        var returnTimeOfPush = 0L
+        when(mockPushOrPullService.send(refEq(NotificationWorkItemWithMetricsTime1), ameq(ApiSubscriptionFieldsOneForPush))(any[HasId])).thenReturn(
+          Future{
+            Thread.sleep(1000)
+            returnTimeOfPush = System.currentTimeMillis()
+            Right(Pull)
+          })
+
+        val result = service.handleNotification(ValidXML, requestMetaData, ApiSubscriptionFieldsOneForPush)
+
+        await(result) shouldBe true
+        val returnTimeOfSavedToDb = System.currentTimeMillis()
+        eventually(verify(mockNotificationWorkItemRepo).permanentlyFailedByClientIdExists(NotificationWorkItemWithMetricsTime1.clientId))
+        eventually(verify(mockNotificationWorkItemRepo).saveWithLock(refEq(NotificationWorkItemWithMetricsTime1), refEq(InProgress)))
+        eventually(verify(mockNotificationWorkItemRepo, times(0)).setCompletedStatus(any[BSONObjectID], any[ResultStatus]))
+        infoLogVerifier("Existing permanently failed notifications found for client id: ClientId. Setting notification to permanently failed")
```

Comment:
rather than `Save to database did not happen before push` how about other way round `Push happened asynchronously after save to database` ?